operator/internal/controller/pulumi/stack_controller.go

```diff
@@ -654,7 +632,16 @@ func (r *StackReconciler) Reconcile(ctx context.Context, request ctrl.Request) (
 	}
 
 	// Step 2. If there are extra environment variables, read them in now and use them for subsequent commands.
-
+	err = sess.setupWorkspace(ctx)
+	if err != nil {
+		if isStalledError(err) {
+			// TODO: why assume it is a cross-namespace error?
+			instance.Status.MarkStalledCondition(pulumiv1.StalledCrossNamespaceRefForbiddenReason, err.Error())
+			return reconcile.Result{}, saveStatus()
+		}
```

Comment:
I think this is because `SetEnvRefsForWorkspace` can return `errNamespaceIsolation` and the only other stall error is `errOtherThanOneSourceSpecified` which is checked outside of `setupWorkspace`.